src/librustdoc/html/render/mod.rs

```diff
@@ -99,29 +97,28 @@ crate fn ensure_trailing_slash(v: &str) -> impl fmt::Display + '_ {
 /// It is intended that this context is a lightweight object which can be fairly
 /// easily cloned because it is cloned per work-job (about once per item in the
 /// rustdoc tree).
-#[derive(Clone)]
 crate struct Context<'tcx> {
     /// Current hierarchy of components leading down to what's currently being
     /// rendered
-    crate current: Vec<String>,
+    current: Vec<String>,
     /// The current destination folder of where HTML artifacts should be placed.
     /// This changes as the context descends into the module hierarchy.
-    crate dst: PathBuf,
+    dst: PathBuf,
     /// A flag, which when `true`, will render pages which redirect to the
     /// real location of an item. This is used to allow external links to
     /// publicly reused items to redirect to the right location.
-    crate render_redirect_pages: bool,
+    render_redirect_pages: bool,
     /// The map used to ensure all generated 'id=' attributes are unique.
-    id_map: Rc<RefCell<IdMap>>,
+    id_map: RefCell<IdMap>,
```

Comment:
I think removing the `Rc` and `Arc` is having much more side-effects that you might expect. :)

Reply:
Right, I have the same comment as before: this is a change in behavior and should probably be dropped.

Reply:
@jyn514 Still think so?

Reply:
The diff is confusing: That change is old, but the new change from #82424 prevents it from being "outdated" from GitHub's point of view.

Before I was trying to get rid of the cloning of `Context` and get rid of the shared state in `Context`, so I removed the Rc and RefCell. Now the code still clones `Context`, but all the shared mutable state is moved to `SharedContext`, and `id_map` isn't shared.

`id_map` wasn't shared before, either. Quoting from f8bd38d:

> I'm not sure why they were shared before. I think the reason `id_map`
worked as a shared value before is that it is cleared before rendering
each item (in `render_item`). [...]
>
> Note that `id_map` currently still has to be cleared because otherwise
child items would inherit the `id_map` of their parent. I'm hoping to
figure out a way to stop cloning `Context`, but until then we have to
reset `id_map`.

I think this thread can be marked as resolved.

Reply:
> Now the code still clones Context, but all the shared mutable state is moved to SharedContext, and id_map isn't shared.

I don't quite follow - isn't `id_map` shared because the same `Context` is used everywhere? What do you mean by shared?

Can you link to where `Context` is cloned? It still seems like a change in behavior to me since cloning will no longer update the same `id_map` in both places.

Reply:
`id_map` is cleared (and was cleared before this change as well) before each item is rendered in `render_item`, so it wasn't shared before and it still isn't shared now.

`id_map` is cleared/reset here:

https://github.com/rust-lang/rust/blob/f8bd38d0af4d23f34cf7e6304d8eda1ab4a549b4/src/librustdoc/html/render/mod.rs#L1592-L1596

And `Context` is cloned here:

https://github.com/rust-lang/rust/blob/f8bd38d0af4d23f34cf7e6304d8eda1ab4a549b4/src/librustdoc/formats/renderer.rs#L81

And here:

https://github.com/rust-lang/rust/blob/f8bd38d0af4d23f34cf7e6304d8eda1ab4a549b4/src/librustdoc/formats/renderer.rs#L101

And used for `after_krate` here:

https://github.com/rust-lang/rust/blob/f8bd38d0af4d23f34cf7e6304d8eda1ab4a549b4/src/librustdoc/formats/renderer.rs#L110-L111